Add some notes about FinalizationRegistry semantics (#26991)
mdn/content: files/en-us/web/javascript/reference/global_objects/finalizationregistry/index.md

Correct use of `FinalizationRegistry` takes careful thought, and it's best avoided if possible. It's also important to avoid relying on any specific behaviors not guaranteed by the specification. When, how, and whether garbage collection occurs is down to the implementation of any given JavaScript engine. Any behavior you observe in one engine may be different in another engine, in another version of the same engine, or even in a slightly different situation with the same version of the same engine. Garbage collection is a hard problem that JavaScript engine implementers are constantly refining and improving their solutions to.

> Here are some specific points that the authors of the proposal included in its [explainer document](https://github.com/tc39/proposal-weakrefs/blob/master/README.md):

Reviewer: This loses context.

"the proposal"? What proposal? Previously it was the "WeakRef proposal" or otherwise this would be okay if a proposal was mentioned beforehand, but the document has not.

Author: Errr... Sure. I was thinking that mentioning "WeakRef" on FinalizationRegistry page was jarring. What about:

```suggestion
Here are some specific points that the authors of the original proposal included in its [explainer document](https://github.com/tc39/proposal-weakrefs/blob/master/README.md):
```

Reviewer: > Errr... Sure. I was thinking that mentioning "WeakRef" on FinalizationRegistry page was jarring. What about:

"original x" refers to an earlier version/variant of x. Here that is ...a proposal. *Some* proposal. Neither the article nor quote make it explicit as to what proposal that is.

I would say "original WeakRef proposal" makes that clear.

How about
```suggestion
Here are some specific points that were included by the authors in the [explainer document](https://github.com/tc39/proposal-weakrefs/blob/master/README.md) of the proposal that introduced `FinalizationRegistry`:
```
Maybe that can be considered wordy, but that's just how *I* write, feel free to condense it apply your style.

> > [Garbage collectors](<https://en.wikipedia.org/wiki/Garbage_collection_(computer_science)>) are complicated. If an application or library depends on GC cleaning up a WeakRef or calling a finalizer \[cleanup callback] in a timely, predictable manner, it's likely to be disappointed: the cleanup may happen much later than expected, or not at all. Sources of variability include:

Reviewer: What is the intent of this change?
As I see it, neither are really relevant.

If the JavaScript runtime/engine collects a `FinalizationRegistry` before all of the objects that are registered with it are cleaned up, this is allowed an observable, as those registered objects will now no-longer have the user `FinalizationRegistry` callback function called, henceforth.

If a `WeakRef` is GCed, well, unless it's in a `inalizationRegistry` itself, it's unobservable. Presumably this intended to refer to the object that a `WeakRef` references, which is actually just an arbitrary object, and would be more in-line with what the document is saying;

```suggestion
> [Garbage collectors](<https://en.wikipedia.org/wiki/Garbage_collection_(computer_science)>) are complicated. If an application or library depends on GC cleaning up an object or calling a finalizer \[cleanup callback] in a timely, predictable manner, it's likely to be disappointed: the cleanup may happen much later than expected, or not at all. Sources of variability include:
```

Author: @Phosra Originally, I was syncing this paragraph with the WeakRef page. This paragraph is verbatim copied from the WeakRef proposal and I don't think it's valuable to change it beyond necessary. The original text said "WeakRef" and I believe changing it to FinalizationRegistry was overzealous, since "calling a finalizer" already encompasses FinalizationRegistries. "Cleaning up a WeakRef" could in some way mean "setting [[WeakRefTarget]] to empty", because in fact "cleaning up an object" does not necessarily mean that it's WeakRef-oblivious.

Reviewer: Sure, disregard this.
